src/OpenTelemetry/Metrics/MetricReaderExt.cs

```diff
@@ -27,6 +27,7 @@ namespace OpenTelemetry.Metrics
     public abstract partial class MetricReader
     {
         private readonly HashSet<string> metricStreamNames = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+        private readonly Dictionary<InstrumentIdentity, Metric> instrumentIdentityToMetric = new Dictionary<InstrumentIdentity, Metric>();
```

Comment:
We need to clear an entry, upon instrument dispose as well.

Reply:
Turns out there's a bit more to just freeing up an entry in this dictionary. I hadn't fully followed the dispose path, so I didn't notice that when an instrument is disposed the metric is freed up. Now that multiple instruments can modify a single metric we need to free up the metric only when no more instruments reference it.

Working on this...

Reply:
Ah ok, my initial understanding was incorrect. Instruments don't get disposed, meters do, so I think is less complicated than I had thought... https://github.com/open-telemetry/opentelemetry-dotnet/pull/2916/commits/0ff1d224122c2fd494826a15d72384a565126cce